### DISCUSSION TITLE
Feature request: Polyline/Curve representation

### DISCUSSION THREAD

**Discussion by bhouston** [elalish/manifold#168] [Ideas]
Another wishlist item, feel free to ignore.

I think this need a Polyline/curve class.  At a simplistic level engineering models usually are created first in 2D and then "extrude"-ed into 3D shapes which are then CSG'ed together.  You are missing the Polyline/curve class that would create the 2D shapes that are fundamental to CAD.  This feels like a gap.

It would be nice to be able to similarly treat them as smoothable Bezier curves like the tri-meshes so they are equally refineable like the trimeshes.

**Comment by elalish** [comment]
I have this: https://elalish.github.io/manifold/classmanifold_1_1_manifold.html#a8631da9e82e21ccf91a53e016aa18c30

`Polygons` are just a vector of vectors; it could use more methods of its own for sure.

**Comment by bhouston** [comment]
Polygon implies it is a closed loop (whatever the 2D equivalent of a manifold is.)  I think that requiring that on Curves/Polylines to always be closed isn't accurate.  I think you often want to create them as not closed and then you may connect a few together to create one or more polygons as a result.

I think that a 2D drawing is fundamentally a collection of lines, many of which are connected and then from that set of connected lines you can extract a set of closed regions, e.g. polygons.  But the mapping of lines to polygons it not 1:1.

**Reply by elalish** [reply]
I think given the name of this library, we're only concerned with manifold geometry. Agreed that separate lines and constraints may be used to create the polygon cross-sections, but I think that's all at a level above this library. Definitely a useful level though!

**Comment by bhouston** [comment]
The lines are absolutely fundamental because these are usually the primitives in the CAD tool upon which the constraints act.  Thus keeping these in a proper representation matters a lot.

Basically everything in CAD is derived from the line set, both polygons and trimeshes in your case.  Thus you need to have these as a first class representation.
